rfd/XXXXa-scoped-join-tokens.md

```diff
@@ -0,0 +1,236 @@
+scoped tokens.
+
+## Overview
+
+### Planned scope
+
+The scope of work proposed by this RFD is:
+
+- A `ScopedToken` resource type and API that is itself guarded by the new
+  scoped access rules
+- Support for static tokens and provisioning using the "token" join method.
+- Support for configuring an `assigned_scope` on a scoped token which will
+  in turn be assigned to new resources at provisioning time
+- Support for configuring a scoped token with a set of labels that are
+  automatically assigned to a resource at provisioning time
+- Support for defining a maximum number of resources a scoped token can
+  provision
+- New `scoped` variants of the `tctl tokens *` family of sub commands
+
+Considered out of scope for this RFD:
+
+- Token types other than static tokens or join methods other than "token".
+  These will almost certainly be implemented in the future but are not part
+  of the immediately planned work.
+
+### Scoped Tokens
+
+The recently added `ScopedToken` resource will be modified to include fields
+necessary for assigning scopes, labels, and enforcing limited uses.
+Additionally, fields missing from the existing `types.ProvisionTokenV2` will be
+ported over to facilitate existing provisioning semantics.
+
+```diff
+--- a/api/proto/teleport/scopes/joining/v1/token.proto
++++ b/api/proto/teleport/scopes/joining/v1/token.proto
+@@ -16,7 +16,9 @@ syntax = "proto3";
+
+ package teleport.scopes.joining.v1;
+
++import "google/protobuf/timestamp.proto";
+ import "teleport/header/v1/metadata.proto";
++import "teleport/legacy/types/types.proto";
+
+ option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/scopes/joining/v1;joiningv1";
+
+@@ -46,8 +48,28 @@ message ScopedToken {
+
+ // ScopedTokenSpec is the specification of a scoped token.
+ message ScopedTokenSpec {
+-  // AssignedScope is the scope to which this token is assigned.
++  // The scope to which resources provisioned using this token are assigned.
+   string assigned_scope = 1;
+
+-  // TODO(fspmarshall): port relevant token features to scoped tokens.
++  // The list of roles associated with the token. They will be converted
++  // to metadata in the SSH and X509 certificates issued to the user of the
++  // token.
++  repeated string roles = 2;
++
++  // The joining method required in order to use this token.
++  // Supported joining methods for scoped tokens only include 'token'.
++  string join_method = 3;
++
++  // The list of rules applied to resources using this token.
++  // At least one rule must match the resource in order to join.
++  repeated types.TokenRule allow = 4;
++
++  // The set of labels to be automatically assigned to any provisioned resource.
++  map<string, string> assigned_labels = 5;
++
++  // The number of resources that can be provisioned using this token.
++  int32 max_uses = 6;
++
++  // The number of resources that have been provisioned using this token.
++  int32 total_uses = 6;
```

Comment:
`spec` and `metadata` should be settable through IaC without having to manually ignore fields that have changed, a field like `total_uses` should be store in the `status` top level field.

Reply:
Most definitely, good call out :+1: